USB Multi-instance support

finikorg opened the pull request:
This is RFC for proposed way of multi - instance USB class support.

Multi - instance means that the same instances of some class might be included to the code, for example 2 CDC ACM (Serial ports). Before you could include only one instance of the same class.
I have created sample illustrating this possibility with 2 USB serial ports.

finikorg committed:
usb: cdc_acm: Allow to create several CDC ACM devices

Change global device pointer to device list where every device
registers.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: cdc_acm: Correct debug logging

Remove extra newline characters.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Refactor usb_interface_config callback

Add interface parameter to interface configuration callback to be able
to configure several instances of the same class.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: device: Add composite status callback

Add new status callback with usb_cfg_data parameter to be able to
identify instance callback.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: cdc_acm: Use usb_cfg_data from config_info

Use saved data in dev->config->config_info.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: cdc_acm: Add macros for multi instance configuration

Add macros for defining multiple instances of CDC ACM devices.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: cdc_acm: Add menuconfig option for multiple instances

Add menuconfig option to select another instance of CDC ACM device.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
> > This is RFC for proposed way of multi - instance support.
> 
> Multi-instance what?

Multiple USB classes (several CDC ACM devices, etc)

MarkWangChinese commented:
HID class driver APIs have no instance parameter like cdc_acm's "struct device *dev", how HID class driver support multiple HID instances? For example: int hid_int_ep_write(const u8_t *data, u32_t data_len, u32_t *bytes_ret). From my viewpoint, it may be necessary to add the device instance parameter to the HID API, and implement one way to define configurable number of the HID interface/endpoint descriptors because the application determine the instances count.  Other classes drivers have the same problem.

How to support the multiple instances of controller and devce stack? for example: RT1050 has two USB controllers, Zephyr application cannot initialize two device stack instances to use the two controllers at the same time. From my viewpoint, it may be necessary to add the device instance parameter to the device stack API and controller API.

If I am wrong, please let me know.

finikorg commented:
@MarkWangChinese I will check also all other classes and push more commits for HID, etc. For the multi controller case, I need to think more, probably we need controller device model and change stack to use this specific device functions.

pawelzadrozniak commented:
Yep, I'd like all implementations of classes to be unified - i.e. defined as a device, all local data (control block) grouped in one structure which could be retrieved with device structure pointer and no other static variables (unless it's a common variable for all instances). 

A couple of ideas.
Macro for retrieving specific class instance count + sum of all instances of any class would be useful. With such macros we could get rid of composite device definition in kconfig.

Each instance will also need a configurable endpoint numbers (kconfig?) and I'd add a test routine executed during usb initialization (included only in debug build) which would go through all instances and check if user has not overlapped EPs in class configurations.

finikorg committed:
usb: hid: Add configuration options for second HID instance

Add configuration option for enabling second HID Device

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: hid: Add HID Device concept

Add HID Device associated with the instance of the HID. This allows to
create several HID instances for multifunction composite device.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: hid: Use HID Device interface for samples

Use new HID Device interface for HID sample

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: hid-mouse: Refactor and use new HID Device interface

Refactor of hid-mouse and using new HID Device interface.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: hid: Use HID Device interface

Use new HID Device interface for tests/boards/intel_s1000_crb

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: fxos8700-hid: Use HID Device interface

Refactor sample using new HID Device interface.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented on samples/subsys/usb/hid-mouse/src/main.c:
```diff
@@ -228,8 +240,9 @@ void main(void)
-	usb_hid_init();
+	usb_hid_register_device(hid_dev, hid_report_desc,
+				sizeof(hid_report_desc), NULL);
+	usb_hid_init(hid_dev);
```
ACM CDC is different from the other classes since it has (serial) device already, all other classes should have device similar to HID Device.

About `usb_hid_init()`: if we move general init to `usb_init()` then the remaining part we can move to `usb_hid_register_device()`

finikorg committed:
samples: usb: Update samples for new port name

Use new multi instance interface and new port names.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: cdc_acm: Use u8_t for interface number

Change argument for get_dev_data_by_iface().

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: hid: Add get_dev_data_by_cfg helper

Reduce time by using the helper.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: hid: Add sys_le16_to_cpu() conversion

Add conversion since interface number is stored in lower byte of
wIndex.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Add helpers for getting common device data

Add helpers to be used in USB classes for getting device data.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: hid: Use new device data interface

Use new interface for getting device data.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: cdc_acm: Use new device data interface

Use device data interface to handle device data.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: cdc_acm: Add composite CDC ACM sample

Add sample creating 2 serial USB ports and establishing communication
between those 2 ports.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

carlescufi commented:
@masz-nordic @pawelzadrozniak @jfischer-phytec-iot Can you please revisit this PR?

finikorg committed:
samples: cdc_acm_composite: Add README documentation

Add README describing the sample.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

nashif reviewed (changes requested):
Please add a body to the PR, explain what that is. Not everyone knows what usb multi-instance means and going thru the commits does not help.

dbkinder commented on samples/subsys/usb/cdc_acm_composite/README.rst:
```diff
@@ -0,0 +1,63 @@
+   Baudrate detected: 115200
+   Baudrate detected: 115200
+
+And on ttyACM devices provided by zephyr USB device stack:
```
```suggestion
And on ttyACM devices provided by the Zephyr USB device stack:
```

dbkinder commented on samples/subsys/usb/cdc_acm_composite/README.rst:
```diff
@@ -0,0 +1,63 @@
+
+   Wait for DTR
+
+Open a serial port emulator, for example minicom
```
```suggestion
Open a serial port emulator, for example, minicom,
```

dbkinder reviewed (commented):
a few grammar and spelling edits

dbkinder reviewed (dismissed):
A few grammar and spelling fixes

finikorg commented:
> Please add a body to the PR, explain what that is. Not everyone knows what usb multi-instance means and going thru the commits does not help.

Updated PR description.

finikorg committed:
usb: cdc_acm: Refactor Kconfig for CDC ACM

Remove unneeded "depends on" and use "if USB_CDC_ACM" instead of.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: msc: Fix redeclaration of enumerators

Add prefixes to MSC enumerators, otherwise they are (ERROR) are
conflicting with other enumerators.

...
subsys/usb/class/mass_storage.c:149:2: error: redeclaration of
enumerator ‘ERROR’
 ERROR,        /* error */
 ^~~~~
...
ext/hal/st/stm32cube/stm32f4xx/soc/stm32f4xx.h:216:3: note: previous
 definition of ‘ERROR’ was here
   ERROR = 0U,
   ^~~~~
...

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jfischer-phytec-iot commented on subsys/usb/class/Kconfig:
```diff
@@ -15,13 +15,29 @@ config USB_CDC_ACM
+config CDC_ACM_PORT_1
+	bool "Enable virtual port 1"
+	help
+	  Enable CDC ACM Virtual port 1
```
It is not virtual, where does it come from? Also "class device driver port name" sounds strange, why not "CDC ACM class name" or "CDC ACM interface name"?

finikorg replied:
@jfischer-phytec-iot `interface` and `class` might be misunderstood in USB. What about 'CDC_ACM_DEVICE_1`
Also another question: Is this OK to have number of instances defined in menuconfig?

finikorg replied:
Just wondering if we specify only number of instances, can we create number of devices with macros in `include/misc/util.h`

MarkWangChinese commented on subsys/usb/usb_device.c:
```diff
@@ -1377,8 +1377,10 @@ static void forward_status_cb(enum usb_dc_status_code status, const u8_t *param)
+		struct usb_cfg_data *cfg = &__usb_data_start[i];
+
+		if (cfg->cb_usb_status_composite) {
+			cfg->cb_usb_status_composite(cfg, status, param);
```
From my understanding, it need calls cfg->cb_usb_status too as follow if cb_usb_status_composite is NULL and cb_usb_status is not NULL.
		if (cfg->cb_usb_status_composite) {
			cfg->cb_usb_status_composite(cfg, status, param);
		} else if (cfg->cb_usb_status) {
			cfg->cb_usb_status(status, param);
		}
For example: mass1 + cdc_acm1 + cdc_acm2, the mass storage class driver use cb_usb_status and cdc class driver use cb_usb_status_composite.
Or we can remove cb_usb_status and replace cb_usb_status with cb_usb_status_composite, and rename cb_usb_status_composite as another name.

Please let me know if I am wrong.

finikorg replied:
For composite configuration we need to add `cb_usb_status_composite` to MSC. Even if now only one instance of MSC can be enabled.

Probably renaming callback would be the right approach. So we might have common callback with `NULL` parameter for non composite configuration in place of `cfg`.

MarkWangChinese replied:
Yes, it is better to rename the callback.
Do you plan to add device instance (struct device *dev ) to all USB class drivers? for example: MSC.
If you plan add it, is it better to add one (struct device *dev) parameter to the follow callback functions?
usb_request_handler class_handler;
usb_request_handler vendor_handler;
usb_request_handler custom_handler;
usb_dc_status_callback cb_usb_status;
void (*cb_usb_status_composite)(struct usb_cfg_data *cfg,
					enum usb_dc_status_code cb_status,
					const u8_t *param);
And add the device_name (const char *dev_name) to struct usb_cfg_data.
After these changes, the usb_get_dev_data_by_cfg, usb_get_dev_data_by_iface and usb_get_dev_data_by_ep are not needed any more.

finikorg replied:
@MarkWangChinese fc34152a3bc7cfa7e668d2bd4524d9727725fe99 renames callbacks. There is other PR  #13310. You can comment there, since this PR is merged already.

MarkWangChinese commented on include/usb/usb_device.h:
```diff
@@ -190,6 +191,11 @@ struct usb_cfg_data {
+	void (*cb_usb_status_composite)(struct usb_cfg_data *cfg,
+					enum usb_dc_status_code cb_status,
+					const u8_t *param);
+
```
#ifdef CONFIG_USB_COMPOSITE_DEVICE may be needed for the above added lines.

finikorg replied:
Will fix it in the follow up patches
